### DISCUSSION TITLE
trivial:  Error Message 'Function with declared return type...' for inferred return type

### DISCUSSION THREAD

**Discussion by pinterior** [microsoft/pyright#8421] [Q&A] [answered]
currently, with following conditions all met, error message `Function with declared return type "NoReturn" cannot include a return statement  (reportGeneralTypeIssues)` shown:

1. the function is a never ending generator function without explicit return type annotation
2. the function returns value from never taken case statement
3. the function is wrapped by return-type-generic wrapper function with `@decorator` syntax

the existence of the error message is useful, but `with declared return type "NoReturn"` part is a bit confusing.
could it be changed to `inferred` or something similar in such a situation?

### simplified example

[on pyright-play.net](https://pyright-play.net/?locale=en&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkAoFgEyWCgHcQCKAbQAqAXQAUwAFz1G8ZkkGDRAGihiAlFJZRdUEElQBXEFijA2AAT4COXc%2BK069AT3RJ47KAGZnuiASoMAAWUACs2npRsAQ4SFAARMBgYAmR0RkGxqaJRAQgCX7RMLHxAPrpGVE8wejI6iBGSJVVGRSxOCxAA)

```python
from collections.abc import Callable

def wrap[T](f: Callable[[], T]):
    return f

@wrap
def f():
    yield 3
    match 5:
        case "foo":
            return "bar"  # Function with declared return type "NoReturn" cannot include a return statement  (reportGeneralTypeIssues)
        case _:
            while True:
                pass
```

**Comment by erictraut** [comment, marked as answer]
Yes, this is a bug. It will be addressed in the next release.
